core/types/receipt.go

```diff
@@ -60,9 +60,10 @@ type Receipt struct {
 
 	// Implementation fields: These fields are added by geth when processing a transaction.
 	// They are stored in the chain database.
-	TxHash          common.Hash    `json:"transactionHash" gencodec:"required"`
-	ContractAddress common.Address `json:"contractAddress"`
-	GasUsed         uint64         `json:"gasUsed" gencodec:"required"`
+	TxHash            common.Hash    `json:"transactionHash" gencodec:"required"`
+	ContractAddress   common.Address `json:"contractAddress"`
+	GasUsed           uint64         `json:"gasUsed" gencodec:"required"`
+	EffectiveGasPrice *big.Int       `json:"effectiveGasPrice,omitempty"`
```

Comment:
Where is the value being set?

Reply:
Inside `UnmarshalJSON` in the file `gen_receipt_json.go`

Reply:
He means within execution. At some point, probably during state processing, you need to set the effective gas price. The way it is right now, effective gas is never set, so it's 0 for all receipts

Reply:
I think it's set in `internal/ethapi/api.go:1641` in the function GetTransactionReceipt.
Personally I checked with my Node (Quicknode) that is running on Geth 1.10.26 and it's perfectly work.

Reply:
I see you add this field so you can use it in `ethclient.TransactionReceipt()`

Reply:
Exactly

Reply:
My first reaction is that these types are foremost used in geth core and we shouldn't change them to satisfy API features. That said there is precedence: the `Log.Removed` field is not used in core processing but only in the API.

@fjl Wdyt?

Reply:
Honestly I don't really know the bests practices on this project since I use it only as library to interacting with an already existing node. But for my use case having this information inside the receipt is an absolute necessity.

Also geth as client is already sending this info in the eth_getTransactionReceipt method but is currently not able to read the info the other side when used as a library, so it's sound like an omission from my point of view.

Reply:
We can add this, if we also set it during receipt derivation.

At the moment, we create the value of this field in the RPC handler: https://github.com/ethereum/go-ethereum/blob/c1aa1db69e74c71f251fc83cf7c120b4d0222728/internal/ethapi/api.go#L1640-L1650

If we had `EffectiveGasPrice` as a field in `Receipt`, I'd want it to be derived from the block/tx along with other fields in `DeriveFields`: https://github.com/ethereum/go-ethereum/blob/c1aa1db69e74c71f251fc83cf7c120b4d0222728/core/types/receipt.go#L399

Reply:
Make sense,

On my understanding we need to know the block header for deriving this field:
If the transaction type is 2 (EIP-1559), the effective gas price will be `header.BaseFee + tx.GasTipCap()`.
For any other transaction type, effective gas price will just be `tx.GasPrice()`.
Is it true ?

The problem is that I don't know the best way to getting the block header in the DeriveFields function, is someone got an idea for this ? Or is there another way to get the effective gas price ?

Reply:
I guess we will just have to pass in either the header, or the basefee, as a parameter into `DeriveFields`.

Reply:
Ok, I will try to doing this this or next week.
Do you agree that if we pass the block header as parameter we can remove `hash` & `number` params ?

Reply:
Yes, they can be removed in that case. Note: I haven't checked myself if the header is easily available at all call sites of DeriveFields. If it is, we're good. Otherwise we'll need to consider it further.

Reply:
Ok I think everything is done, I have just a little concern about the comment I made to `accessors_chain.go` that you can see below